src/System.Device.Gpio/System/Device/Gpio/Drivers/LibGpiodDriver.Linux.cs

```diff
@@ -0,0 +1,276 @@
+
+        protected internal override int PinCount => Interop.GetNumberOfLines(_chip);
+
+        public LibGpiodDriver()
+        {
+            OpenChip();
+            _pinNumberToSafeLineHandle = new Dictionary<int, SafeLineHandle>(PinCount);
+        }
+
+        private void OpenChip()
+        {
+            SafeChipIteratorHandle iterator = Interop.GetChipIterator();
+            if (iterator == null)
+            {
+                throw new IOException($"No any chip available, error code: {Marshal.GetLastWin32Error()}");
+            }
+            _chip = Interop.GetNextChipFromChipIterator(iterator);
+            if (_chip == null)
+            {
+                throw new IOException($"No chip found, error code: {Marshal.GetLastWin32Error()}");
+            }
+            // Freeing other chips opened
+            Interop.FreeChipIteratorNoCloseCurrentChip(iterator);
+        }
+
+        private SafeLineHandle RequestEvent(int pinNumber, PinEventTypes eventType)
+        {
+            int eventSuccess = -1;
+            SafeLineHandle pin = Interop.GetChipLineByOffset(_chip, pinNumber);
+            if (pin != null)
+            {
+                if (eventType == PinEventTypes.Rising || eventType == PinEventTypes.Falling)
+                {
+                    eventSuccess = Interop.RequestBothEdgeEventForLine(pin, $"Listen {pinNumber} for falling edge event");
+                }
+                else
+                {
+                    throw new ArgumentException("Invalid or not supported event type requested");
+                }
+            }
+            if (eventSuccess < 0)
+            {
+                throw new IOException($"Error while requesting event listener for pin {pinNumber}, error code: {Marshal.GetLastWin32Error()}");
+            }
+            return pin;
+        }
+
+        protected internal override void AddCallbackForPinValueChangedEvent(int pinNumber, PinEventTypes eventType, PinChangeEventHandler callback)
+        {
+            if (!_pinNumberToEventHandler.TryGetValue(pinNumber, out LibGpiodDriverEventHandler eventHandler))
+            {
+                if (_pinNumberToSafeLineHandle.TryGetValue(pinNumber, out SafeLineHandle pin))
+                    Interop.ReleaseGpiodLine(pin);
+                eventHandler = new LibGpiodDriverEventHandler(pinNumber, new CancellationTokenSource());
+                _pinNumberToEventHandler.TryAdd(pinNumber, eventHandler);
+                eventHandler.PinHandle = RequestEvent(pinNumber, eventType);
+                InitializeEventDetectionTask(eventHandler);
+            }
+            if (eventType == PinEventTypes.Rising)
+            {
+                eventHandler.ValueRising += callback;
+            }
+            else if (eventType == PinEventTypes.Falling)
+            {
+                eventHandler.ValueFalling += callback;
+            }
+        }
+
+        private void InitializeEventDetectionTask(LibGpiodDriverEventHandler eventHandler)
```

Comment:
Shouldn't you return the Task that way you can cancel it or make sure it won't continue running when the object is disposed?

Reply:
not sure how to cancel it from Task, as in sys fs driver used CancellationTokenSource for cancelling this thread. But forgot about disposal, updated dispose method to cancel CancellationToken too